### PR TITLE
Fix is_subpath

### LINKED ISSUE #3627: Question: How to run system & idempotence tests
Not sure if this is a user error on my part or a bug/issue with these tests not running.

When I run `cargo test` in my local env, the system and idempotence tests don't appear to be running. Relevant snippets from the output:
```sh
Ran 0 idempotent tests.
test test::idempotence_tests ... ok
.....
Ran 0 system tests.
test test::system_tests ... ok
```

This is also the case in the latest builds on Travis and AppVeyor. For example 
https://travis-ci.com/rust-lang/rustfmt/jobs/207595534
https://ci.appveyor.com/project/rust-lang-libs/rustfmt/build/job/il5kok93xmvcaafu
Relevant snippets from the output:
```sh
Ran 0 idempotent tests.
test test::idempotence_tests ... ok
.....
Ran 0 system tests.
test test::system_tests ... ok
```

However, it seems these tests only started being skipped in Travis and AppVeyor ~1 week ago. Before, it seems that both the system and idempotence tests were always running on each job in both Travis and AppVeyor.

For example 
https://travis-ci.com/rust-lang/rustfmt/jobs/206494618
https://ci.appveyor.com/project/rust-lang-libs/rustfmt/builds/25135606/job/fhkd2ac6bin3x88p
Relevant snippets from the output:
```sh
Ran 473 idempotent tests.
test test::idempotence_tests ... ok
...
Ran 372 system tests.
test test::system_tests ... ok
```

Based on the [contributing guide](https://github.com/rust-lang/rustfmt/blob/master/Contributing.md#create-test-cases) I would have expected both the system and idempotence tests to be executed as part of `cargo test`. Is that no longer the case or is there a bug in the discovery/execution of those tests?

### PR DESCRIPTION
**Pull request by topecongiro**
Closes #3627.

### CODE REVIEW

File: src/test/mod.rs [hunk 1]
```diff
@@ -65,9 +65,21 @@ fn is_subpath<P>(path: &Path, subpath: &P) -> bool
...
+        I: Iterator<Item = J>,
+        J: Iterator<Item = std::path::Component<'a>>,
+        F: Fn() -> T,
+        T: Iterator<Item = std::path::Component<'a>>,
+    {
+        paths.any(|c| c.zip(subpath_components()).all(|(a, b)| a == b))
+    }
+
+    let range = 1..=path.components().count();
+    let prefixes = range.clone().map(|i| path.components().take(i));
```
Comment (sinkuu) [inline src/test/mod.rs]: If the first prefix doesn't match (`[foo]/bar != [bar]/baz`), then incorporating more components is useless (`[foo/bar] != [bar/baz]`), I guess?

Comment (topecongiro) [reply src/test/mod.rs]: Yeah, you are right. I think what @calebcartwright did in https://github.com/rust-lang/rustfmt/issues/3627#issuecomment-501970476 is better than what I did first, so I have updated this PR with @calebcartwright's snippet.

**Comment by scampi** [comment]
thanks!

**Comment by calebcartwright** [comment]
Thank you! And glad I was able to be of some assistance 👍 

I was also thinking it would be good to augment the existing tests to catch this "no system/idempotence tests ran" scenario to prevent a future recurrence. The existing test cases include a check to ensure there were 0 failed tests, so perhaps something as simple as adding another check that ensures some minimum number of tests ran/passed could address that.

Do folks think that would be valuable/helpful? If so I'd be happy to submit a PR

**Comment by scampi** [comment]
@calebcartwright that'd be nice to have, to catch such things in the future!
